src/Tracking/DisasterTrackerApp.BL/Implementation/UsersGoogleCalendarDataUpdatingService.cs

```diff
@@ -0,0 +1,182 @@
+    {
+        var watchData = WatchChannelDataMapper.MapChannelDataEntityToDto(_watchChannelsRepository.GetWatchChannel(channelToken));
+        if (watchData == null)
+        {
+            _logger.LogWarning("Cannot find watch channel by token {Token}", channelToken);
+            return;
+        }
+        
+        var calendar = (await _calendarRepository.GetFilteredAsync(c => c.UserId == watchData.UserId))
+                        .FirstOrDefault();
+        
+        if (calendar == null)
+        {
+            _logger.LogWarning("Cannot find primary calendar for user {Id}" +
+                               "Unable to process watched resource notification. Token: {Token}", 
+                      watchData.UserId.ToString(), channelToken);
+            return;
+        }
+        
+        var events = await _calendarService.GetUserEventsAsync(watchData.UserId, calendar.GoogleCalendarId, watchData.LastTimeTriggered);
+        await UpdateCalendarEvents(calendar.Id, events);
+        
+        _watchChannelsRepository.Save(channelToken, WatchChannelDataMapper.MapChannelDataDtoToEntity(watchData.UpdateTriggerTime(triggerTime)));
+    }
+
+    private async Task<GoogleCalendar?> UpdateUserCalendar(Guid userId)
+    {
+        var primaryCalendar = await _calendarService.GetUserCalendarAsync(userId);
+        if (primaryCalendar == null)
+        {
+            _logger.LogWarning("Cannot find primary calendar for user {Id}", userId.ToString());
+            
+            return null;
+        }
+        
+        var existingUserCalendar = (await _calendarRepository.GetFilteredAsync(
+            c => c.GoogleCalendarId == primaryCalendar.GoogleCalendarId)).FirstOrDefault();
+
+        if (existingUserCalendar != null)
+        {
+            await _calendarRepository.UpdateAsync(existingUserCalendar, primaryCalendar);
+            return existingUserCalendar;
+        }
+        
+        await _calendarRepository.SaveAsync(primaryCalendar);
+        
+        return primaryCalendar;
+    }
+
+    private async Task<bool> UpdateCalendarEvents(Guid calendarId, IEnumerable<Event> calendarEvents)
+    {
+        try
+        {
+            var changedCalendarEvents = calendarEvents.ToList();
+            var existingCalendarEvents = (await _calendarEventsRepository.GetFilteredAsync(
+                    e => e.CalendarId == calendarId))
+                .ToList();
+
+            var eventsToDelete = changedCalendarEvents
+                .Select(e => e.Id)
+                .Intersect(existingCalendarEvents.Select(e => e.GoogleEventId))
+                .Select(id => existingCalendarEvents.First(e => e.GoogleEventId == id));
+
+            await _calendarEventsRepository.RemoveUserEventsAsync(eventsToDelete);
+
+            await RemoveOutdatedCalendarEvents(existingCalendarEvents);
+
+            var updatedEvents = await Task.WhenAll(changedCalendarEvents
```

Comment:
![image](https://user-images.githubusercontent.com/46414904/148643939-e022e3e1-d00e-4ec6-b7ff-296c84406185.png)

Reply:
Fair point. 
An improvement point for future versions.
Share link please